### ISSUE TITLE
Flannel connectivity seems to be broken in certain situations

### ISSUE THREAD

**Issue by celskeggs** [sipb/homeworld#247]
This comes up occasionally, where a deployment will just happen to fail such that prometheus can't use the `eggs-benedict` (master01) node to connect to the flannel-monitor-collector service.

I suspect that this connectivity problem occurs exactly when a node has been assigned the 172.18.0.0/24 subnet (out of the 172.18.0.0/16 address space) and the connection is started from the node itself, rather than from a container running on the node.

We would only notice this when the node in question happens to be master01, because prometheus routes through master01 by default in the current configuration.

This needs to be checked by redeploying a cluster, and directly attempting to curl various flannel endpoints from each node. In the current cluster, direct access from the node to the cluster works on all nodes except master01, which is currently assigned 172.18.0.0/24. My hypothesis is that whichever node is assigned 172.18.0.0/24 will fail to access the rest of the flannel network.

**Comment by celskeggs** [comment]
Confirmed. This is, in fact, the situation in which this problem appears.

**Comment by celskeggs** [comment]
Discovery: connections on the affected node work when the kvm-cni interface is directly specified.

Interface list:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:00:35:eb:cc brd ff:ff:ff:ff:ff:ff
    inet 18.181.0.129/16 brd 18.181.255.255 scope global ens3
       valid_lft forever preferred_lft forever
    inet6 2001:4830:2446:b5:5054:ff:fe35:ebcc/64 scope global mngtmpaddr dynamic 
       valid_lft 2591685sec preferred_lft 604485sec
    inet6 fe80::5054:ff:fe35:ebcc/64 scope link 
       valid_lft forever preferred_lft forever
8: flannel0: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1472 qdisc pfifo_fast state UNKNOWN group default qlen 500
    link/none 
    inet 172.18.0.0/16 scope global flannel0
       valid_lft forever preferred_lft forever
    inet6 fe80::eba4:f181:b0ac:82dd/64 scope link flags 800 
       valid_lft forever preferred_lft forever
9: kvm-cni0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 00:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/24 scope global kvm-cni0
       valid_lft forever preferred_lft forever
    inet6 fe80::42d:a8ff:fed5:9e04/64 scope link 
       valid_lft forever preferred_lft forever
```

Only this invocation worked:

    curl --interface 172.18.0.1 172.18.5.2

All other interfaces besides 172.18.0.1 failed, whereas all of them worked on other nodes.

**Comment by celskeggs** [comment]
It appears that the distinction is that there are two IP addresses that could be used as the source of a connection from the host: 172.18.x.0 and 172.18.x.1. Connections from 172.18.0.0 are dropped in all cases. When a request is made from an interface besides kvm-cni0 or flannel0, iptables rules are used to reroute it so that the connection comes through 172.18.x.0.

It seems that outbound traffic (to the flannel0 interface) appears in iftop in both cases, but will receive no response when it comes from 172.18.0.0. It is not yet clear whether the traffic is lost on the local side of the connection or the remote side of the connection.

**Comment by celskeggs** [comment]
Here's the [relevant diagram](https://wiki.archlinux.org/index.php/Iptables#Basic_concepts):

```
                               XXXXXXXXXXXXXXXXXX
                             XXX     Network    XXX
                               XXXXXXXXXXXXXXXXXX
                                       +
                                       |
                                       v
 +-------------+              +------------------+
 |table: filter| <---+        | table: nat       |
 |chain: INPUT |     |        | chain: PREROUTING|
 +-----+-------+     |        +--------+---------+
       |             |                 |
       v             |                 v
 [local process]     |           ****************          +--------------+
       |             +---------+ Routing decision +------> |table: filter |
       v                         ****************          |chain: FORWARD|
****************                                           +------+-------+
Routing decision                                                  |
****************                                                  |
       |                                                          |
       v                        ****************                  |
+-------------+       +------>  Routing decision  <---------------+
|table: nat   |       |         ****************
|chain: OUTPUT|       |               +
+-----+-------+       |               |
      |               |               v
      v               |      +-------------------+
+--------------+      |      | table: nat        |
|table: filter | +----+      | chain: POSTROUTING|
|chain: OUTPUT |             +--------+----------+
+--------------+                      |
                                      v
                               XXXXXXXXXXXXXXXXXX
                             XXX    Network     XXX
                               XXXXXXXXXXXXXXXXXX
```

By looking at the output of `iptables -L -v -n`, it appears that connections that get routed down the FORWARD chain succeed, whereas connections that get routed down the INPUT chain fail. My hypothesis is that the main problem occurring is that this choice is made incorrectly in the case of messages from 172.18.0.0.

I don't have enough knowledge about the linux networking model to understand how this routing choice is made. It might be the case that there's something fundamentally wrong about a `x.y.0.0` IP address, or it might be a case where some rule somewhere should be modified for this to work properly.

**Comment by celskeggs** [comment]
It appears that this problem is caused by the following rule in the `local` routing table:

    broadcast 172.18.0.0 dev flannel0 proto kernel scope link src 172.18.5.0

Deleting it fixes the problem, but this isn't a good solution, because it would need to be repeated at least on every restart, if not every time that just the flannel pod restarts.

It appears that [they think they fixed the problem upstream](https://github.com/coreos/flannel/issues/533), but it looks like their fix only applies to the `vxlan` backend.

Solutions to try:

 - Can we patch flannel to use 172.18.x.1 instead of 172.18.x.0?
 - Can we patch kubernetes to not hand out the zero subnet?
 - Can we patch flannel to delete the broadcast route?
 - Can we report this bug to upstream and get them to figure out the *correct* way to fix it?
 - Can we switch to a different backend (like vxlan) where this flaw has been fixed?

**Comment by celskeggs** [comment]
Best solution, as per upstream: change the bound address to be a /32 instead of a /16. A small patch to flannel fixes the problem!

I'll work on getting the patch into upstream.

**Comment by celskeggs** [comment]
[Patch merged upstream.](https://github.com/coreos/flannel/pull/941) We've also patched this locally, which will be PR'd once #246 is merged.
